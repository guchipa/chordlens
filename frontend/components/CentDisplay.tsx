"use client";

import React from "react";
import { Card, CardContent, CardHeader, CardTitle } from "@/components/ui/card";
import { formType } from "@/lib/schema";
import { getEqualJustDiff } from "@/lib/audio_analysis/calcJustFreq";

import { PITCH_COLOR_MAP } from "@/lib/constants";

// 表示するcentの小数点以下桁数
const CENT_DIGIT_NUM = 2;

interface CentDisplayProps {
  pitchList: formType[];
  a4Freq: number;
  title?: string;
}

export const CentDisplay: React.FC<CentDisplayProps> = ({ pitchList, a4Freq, title }) => {
  const equalJustDiff = getEqualJustDiff(pitchList, a4Freq);

  return (
    <Card className="w-full max-w-md mx-auto">
      <CardHeader>
        <CardTitle>{title}</CardTitle>
      </CardHeader>
      <CardContent>
        <div className="flex flex-wrap justify-start gap-x-4 gap-y-2 mt-4">
          {pitchList.map(({ pitchName, octaveNum }, index) => (
            <div
              key={`${pitchName}-${octaveNum}-${index}`}
              className="flex items-center gap-2 text-sm"
            >
              <span
                className="w-3 h-3 rounded-full"
                style={{
                  backgroundColor: PITCH_COLOR_MAP[pitchName] || "#888",
                }}
              />
              <span className="font-semibold">
                {pitchName}
                {octaveNum}
              </span>
              <span>
                {equalJustDiff[index] !== undefined
                  ? `${equalJustDiff[index].toFixed(CENT_DIGIT_NUM)} cents`
                  : "---"}
              </span>
            </div>
          ))}
        </div>
      </CardContent>
    </Card>
  );
};
