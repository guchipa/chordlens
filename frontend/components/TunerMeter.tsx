"use client";

import React from "react";
import { Card, CardContent, CardHeader, CardTitle } from "@/components/ui/card";
import { METER_MAX_DEVIATION_DEGREES } from "@/lib/constants";
import { formType } from "@/lib/schema";

// 各音名に対応する色のマップ
const PITCH_COLOR_MAP: { [key: string]: string } = {
  "C": "#ff6b6b",
  "C#": "#ff8e53",
  "D": "#ffc107",
  "E♭": "#fde047",
  "E": "#a8e063",
  "F": "#56ab2f",
  "F#": "#26de81",
  "G": "#2bcbba",
  "G#": "#45aaf2",
  "A": "#0fb9b1",
  "B♭": "#4a90e2",
  "B": "#8e44ad",
};

interface TunerMeterProps {
  /**
   * 解析結果の配列。各要素が1つの針に対応します。
   */
  analysisData: Array<{
    pitch: formType;
    deviation: number | null;
  }>;
  /** メーターのタイトル */
  title?: string;
}

export const TunerMeter: React.FC<TunerMeterProps> = ({ analysisData, title }) => {
  return (
    <Card className="w-full max-w-md mx-auto">
      <CardHeader>
        <CardTitle>{title || "Tuner"}</CardTitle>
      </CardHeader>
      <CardContent className="flex flex-col justify-center items-center">
        <div className="flex w-64 h-32 overflow-hidden mb-4">
          {/* メーターの背景（半円） */}
          <svg
            className="w-full h-full"
            viewBox="0 0 100 50"
          >
            <path d="M 0 50 A 50 50 0 0 1 100 50 L 0 50 Z" fill="#e0e0e0" />
            {/* 目盛り */}
            {[
              -METER_MAX_DEVIATION_DEGREES,
              -METER_MAX_DEVIATION_DEGREES / 2,
              0,
              METER_MAX_DEVIATION_DEGREES / 2,
              METER_MAX_DEVIATION_DEGREES,
            ].map((angle, idx) => {
              const rotationRad = (angle + 90) * (Math.PI / 180);
              const x1 = 50 + 40 * Math.cos(rotationRad);
              const y1 = 50 - 40 * Math.sin(rotationRad); // Y軸の向きを修正
              const x2 = 50 + 45 * Math.cos(rotationRad);
              const y2 = 50 - 45 * Math.sin(rotationRad); // Y軸の向きを修正
              return (
                <line
                  key={idx}
                  x1={x1}
                  y1={y1}
                  x2={x2}
                  y2={y2}
                  stroke="black"
                  strokeWidth="1"
                />
              );
            })}

            {/* 各音に対応する針を描画 */}
            {analysisData.map(({ pitch, deviation }, index) => {
              if (deviation === null) return null; // 検出されなかった音は表示しない

              const needleRotation =
                deviation * METER_MAX_DEVIATION_DEGREES;
              const needleColor =
                PITCH_COLOR_MAP[pitch.pitchName] || "#888"; // マップにない場合はデフォルト色

              return (
                <line
                  key={`${pitch.pitchName}-${pitch.octaveNum}-${index}`}
                  x1="50"
                  y1="50"
                  x2="50"
                  y2="10"
                  stroke={needleColor}
                  strokeWidth="2.5"
                  strokeLinecap="round"
                  style={{
                    transformOrigin: "50px 50px",
                    transform: `rotate(${needleRotation}deg)`,
                    transition: "transform 0.2s ease-out",
                  }}
                />
              );
            })}
            <circle cx="50" cy="50" r="3" fill="black" />
          </svg>
        </div>
      </CardContent>
    </Card>
  );
};
