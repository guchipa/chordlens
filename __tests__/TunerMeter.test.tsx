
import { render, screen } from "@testing-library/react";
import { TunerMeter } from "@/components/TunerMeter";

describe("TunerMeter", () => {
  it("renders correctly with no deviation", () => {
    const { container } = render(
      <TunerMeter deviation={null} pitchName="A4" />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText("A4")).toBeInTheDocument();
    expect(screen.getByText("--.--%")).toBeInTheDocument();
    expect(screen.getByText("No Sound")).toBeInTheDocument();
  });

  it("renders correctly when in tune", () => {
    const { container } = render(
      <TunerMeter deviation={0.05} pitchName="C4" />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText("C4")).toBeInTheDocument();
    expect(screen.getByText("+5.00%")).toBeInTheDocument();
    expect(screen.getByText("In Tune")).toBeInTheDocument();
  });

  it("renders correctly when out of tune", () => {
    const { container } = render(
      <TunerMeter deviation={-0.5} pitchName="G#3" />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText("G#3")).toBeInTheDocument();
    expect(screen.getByText("-50.00%")).toBeInTheDocument();
    expect(screen.getByText("Out of Tune")).toBeInTheDocument();
  });
});
