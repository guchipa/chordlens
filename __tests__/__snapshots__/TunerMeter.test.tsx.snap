// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TunerMeter renders correctly when in tune 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm w-full max-w-md mx-auto"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <h3
        class="text-2xl font-semibold leading-none tracking-tight"
      >
        Tuner
      </h3>
    </div>
    <div
      class="p-6 pt-0 flex flex-col justify-center items-center"
    >
      <div
        class="flex w-64 h-32 overflow-hidden mb-4"
      >
        <svg
          class="w-full h-full"
          viewBox="0 0 100 50"
        >
          <path
            d="M 0 50 A 50 50 0 0 1 100 50 L 0 50 Z"
            fill="#e0e0e0"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="90"
            x2="95"
            y1="50"
            y2="50"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="78.2842712474619"
            x2="81.81980515339464"
            y1="21.7157287525381"
            y2="18.180194846605364"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="50"
            x2="50"
            y1="10"
            y2="5"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="21.7157287525381"
            x2="18.180194846605364"
            y1="21.715728752538098"
            y2="18.18019484660536"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="10"
            x2="5"
            y1="49.99999999999999"
            y2="49.99999999999999"
          />
          <line
            stroke="#ff6b6b"
            stroke-linecap="round"
            stroke-width="2.5"
            style="transform-origin: 50px 50px; transform: rotate(0deg); transition: transform 0.2s ease-out;"
            x1="50"
            x2="50"
            y1="50"
            y2="10"
          />
          <circle
            cx="50"
            cy="50"
            fill="black"
            r="3"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`TunerMeter renders correctly when out of tune 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm w-full max-w-md mx-auto"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <h3
        class="text-2xl font-semibold leading-none tracking-tight"
      >
        Tuner
      </h3>
    </div>
    <div
      class="p-6 pt-0 flex flex-col justify-center items-center"
    >
      <div
        class="flex w-64 h-32 overflow-hidden mb-4"
      >
        <svg
          class="w-full h-full"
          viewBox="0 0 100 50"
        >
          <path
            d="M 0 50 A 50 50 0 0 1 100 50 L 0 50 Z"
            fill="#e0e0e0"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="90"
            x2="95"
            y1="50"
            y2="50"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="78.2842712474619"
            x2="81.81980515339464"
            y1="21.7157287525381"
            y2="18.180194846605364"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="50"
            x2="50"
            y1="10"
            y2="5"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="21.7157287525381"
            x2="18.180194846605364"
            y1="21.715728752538098"
            y2="18.18019484660536"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="10"
            x2="5"
            y1="49.99999999999999"
            y2="49.99999999999999"
          />
          <line
            stroke="#45aaf2"
            stroke-linecap="round"
            stroke-width="2.5"
            style="transform-origin: 50px 50px; transform: rotate(-45deg); transition: transform 0.2s ease-out;"
            x1="50"
            x2="50"
            y1="50"
            y2="10"
          />
          <circle
            cx="50"
            cy="50"
            fill="black"
            r="3"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`TunerMeter renders correctly with no deviation 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm w-full max-w-md mx-auto"
  >
    <div
      class="flex flex-col space-y-1.5 p-6"
    >
      <h3
        class="text-2xl font-semibold leading-none tracking-tight"
      >
        Tuner
      </h3>
    </div>
    <div
      class="p-6 pt-0 flex flex-col justify-center items-center"
    >
      <div
        class="flex w-64 h-32 overflow-hidden mb-4"
      >
        <svg
          class="w-full h-full"
          viewBox="0 0 100 50"
        >
          <path
            d="M 0 50 A 50 50 0 0 1 100 50 L 0 50 Z"
            fill="#e0e0e0"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="90"
            x2="95"
            y1="50"
            y2="50"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="78.2842712474619"
            x2="81.81980515339464"
            y1="21.7157287525381"
            y2="18.180194846605364"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="50"
            x2="50"
            y1="10"
            y2="5"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="21.7157287525381"
            x2="18.180194846605364"
            y1="21.715728752538098"
            y2="18.18019484660536"
          />
          <line
            stroke="black"
            stroke-width="1"
            x1="10"
            x2="5"
            y1="49.99999999999999"
            y2="49.99999999999999"
          />
          <circle
            cx="50"
            cy="50"
            fill="black"
            r="3"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
